Noisy failure if ROCM not installed

On CPU CI AMDGPU is rather noisy when there is no ROCM installed.

```
┌ AMDGPU → AMDGPUChainRulesCoreExt
│  ┌ Error: ROCm discovery failed!
│  │ Discovered ROCm path: .
│  │ Use `ROCM_PATH` env variable to specify ROCm directory.
│  │ 
│  │   exception =
│  │    AssertionError: isdir(libdir)
│  │    Stacktrace:
│  │      [1] find_rocm_library(lib::String; rocm_path::String, ext::String)
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:110
│  │      [2] find_rocm_library
│  │        @ ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:108 [inlined]
│  │      [3] __init__()
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:79
│  │      [4] run_module_init(mod::Module, i::Int64)
│  │        @ Base ./loading.jl:1193
│  │      [5] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
│  │        @ Base ./loading.jl:1181
│  │      [6] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::Nothing, depmods::Vector{Any})
│  │        @ Base ./loading.jl:1126
│  │      [7] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
│  │        @ Base ./loading.jl:1651
│  │      [8] _require(pkg::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2008
│  │      [9] __require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1882
│  │     [10] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [11] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [12] _require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1873
│  │     [13] macro expansion
│  │        @ ./loading.jl:1860 [inlined]
│  │     [14] macro expansion
│  │        @ ./lock.jl:267 [inlined]
│  │     [15] __require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1823
│  │     [16] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [17] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [18] require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1816
│  │     [19] include
│  │        @ ./Base.jl:495 [inlined]
│  │     [20] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
│  │        @ Base ./loading.jl:2292
│  │     [21] top-level scope
│  │        @ stdin:4
│  │     [22] eval
│  │        @ ./boot.jl:385 [inlined]
│  │     [23] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│  │        @ Base ./loading.jl:2146
│  │     [24] include_string
│  │        @ ./loading.jl:2156 [inlined]
│  │     [25] exec_options(opts::Base.JLOptions)
│  │        @ Base ./client.jl:321
│  │     [26] _start()
│  │        @ Base ./client.jl:557
│  └ @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:101
└  
┌ Trixi → TrixiAMDGPUExt
│  ┌ Error: ROCm discovery failed!
│  │ Discovered ROCm path: .
│  │ Use `ROCM_PATH` env variable to specify ROCm directory.
│  │ 
│  │   exception =
│  │    AssertionError: isdir(libdir)
│  │    Stacktrace:
│  │      [1] find_rocm_library(lib::String; rocm_path::String, ext::String)
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:110
│  │      [2] find_rocm_library
│  │        @ ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:108 [inlined]
│  │      [3] __init__()
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:79
│  │      [4] run_module_init(mod::Module, i::Int64)
│  │        @ Base ./loading.jl:1193
│  │      [5] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
│  │        @ Base ./loading.jl:1181
│  │      [6] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::Nothing, depmods::Vector{Any})
│  │        @ Base ./loading.jl:1126
│  │      [7] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
│  │        @ Base ./loading.jl:1651
│  │      [8] _require(pkg::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2008
│  │      [9] __require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1882
│  │     [10] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [11] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [12] _require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1873
│  │     [13] macro expansion
│  │        @ ./loading.jl:1860 [inlined]
│  │     [14] macro expansion
│  │        @ ./lock.jl:267 [inlined]
│  │     [15] __require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1823
│  │     [16] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [17] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [18] require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1816
│  │     [19] include
│  │        @ ./Base.jl:495 [inlined]
│  │     [20] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
│  │        @ Base ./loading.jl:2292
│  │     [21] top-level scope
│  │        @ stdin:4
│  │     [22] eval
│  │        @ ./boot.jl:385 [inlined]
│  │     [23] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│  │        @ Base ./loading.jl:2146
│  │     [24] include_string
│  │        @ ./loading.jl:2156 [inlined]
│  │     [25] exec_options(opts::Base.JLOptions)
│  │        @ Base ./client.jl:321
│  │     [26] _start()
│  │        @ Base ./client.jl:557
│  └ @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:101
└  
┌ MKL_jll
│   Downloading artifact: IntelOpenMP
└  
┌ AMDGPU → AMDGPUEnzymeCoreExt
│  ┌ Error: ROCm discovery failed!
│  │ Discovered ROCm path: .
│  │ Use `ROCM_PATH` env variable to specify ROCm directory.
│  │ 
│  │   exception =
│  │    AssertionError: isdir(libdir)
│  │    Stacktrace:
│  │      [1] find_rocm_library(lib::String; rocm_path::String, ext::String)
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:110
│  │      [2] find_rocm_library
│  │        @ ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:108 [inlined]
│  │      [3] __init__()
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:79
│  │      [4] run_module_init(mod::Module, i::Int64)
│  │        @ Base ./loading.jl:1193
│  │      [5] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
│  │        @ Base ./loading.jl:1181
│  │      [6] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::Nothing, depmods::Vector{Any})
│  │        @ Base ./loading.jl:1126
│  │      [7] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
│  │        @ Base ./loading.jl:1651
│  │      [8] _require(pkg::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2008
│  │      [9] __require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1882
│  │     [10] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [11] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [12] _require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1873
│  │     [13] macro expansion
│  │        @ ./loading.jl:1860 [inlined]
│  │     [14] macro expansion
│  │        @ ./lock.jl:267 [inlined]
│  │     [15] __require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1823
│  │     [16] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [17] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [18] require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1816
│  │     [19] include
│  │        @ ./Base.jl:495 [inlined]
│  │     [20] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
│  │        @ Base ./loading.jl:2292
│  │     [21] top-level scope
│  │        @ stdin:4
│  │     [22] eval
│  │        @ ./boot.jl:385 [inlined]
│  │     [23] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│  │        @ Base ./loading.jl:2146
│  │     [24] include_string
│  │        @ ./loading.jl:2156 [inlined]
│  │     [25] exec_options(opts::Base.JLOptions)
│  │        @ Base ./client.jl:321
│  │     [26] _start()
│  │        @ Base ./client.jl:557
│  └ @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:101
└  
┌ MPI → AMDGPUExt
│  ┌ Error: ROCm discovery failed!
│  │ Discovered ROCm path: .
│  │ Use `ROCM_PATH` env variable to specify ROCm directory.
│  │ 
│  │   exception =
│  │    AssertionError: isdir(libdir)
│  │    Stacktrace:
│  │      [1] find_rocm_library(lib::String; rocm_path::String, ext::String)
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:110
│  │      [2] find_rocm_library
│  │        @ ~/.julia/packages/AMDGPU/maWzW/src/discovery/utils.jl:108 [inlined]
│  │      [3] __init__()
│  │        @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:79
│  │      [4] run_module_init(mod::Module, i::Int64)
│  │        @ Base ./loading.jl:1193
│  │      [5] register_restored_modules(sv::Core.SimpleVector, pkg::Base.PkgId, path::String)
│  │        @ Base ./loading.jl:1181
│  │      [6] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::Nothing, depmods::Vector{Any})
│  │        @ Base ./loading.jl:1126
│  │      [7] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128)
│  │        @ Base ./loading.jl:1651
│  │      [8] _require(pkg::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2008
│  │      [9] __require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1882
│  │     [10] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [11] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [12] _require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:1873
│  │     [13] macro expansion
│  │        @ ./loading.jl:1860 [inlined]
│  │     [14] macro expansion
│  │        @ ./lock.jl:267 [inlined]
│  │     [15] __require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1823
│  │     [16] #invoke_in_world#3
│  │        @ ./essentials.jl:926 [inlined]
│  │     [17] invoke_in_world
│  │        @ ./essentials.jl:923 [inlined]
│  │     [18] require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:1816
│  │     [19] top-level scope
│  │        @ ~/.julia/packages/MPI/TKXAj/ext/AMDGPUExt.jl:4
│  │     [20] include
│  │        @ ./Base.jl:495 [inlined]
│  │     [21] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
│  │        @ Base ./loading.jl:2292
│  │     [22] top-level scope
│  │        @ stdin:4
│  │     [23] eval
│  │        @ ./boot.jl:385 [inlined]
│  │     [24] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│  │        @ Base ./loading.jl:2146
│  │     [25] include_string
│  │        @ ./loading.jl:2156 [inlined]
│  │     [26] exec_options(opts::Base.JLOptions)
│  │        @ Base ./client.jl:321
│  │     [27] _start()
│  │        @ Base ./client.jl:557
│  └ @ AMDGPU.ROCmDiscovery ~/.julia/packages/AMDGPU/maWzW/src/discovery/discovery.jl:101
└  
     Testing Running tests...
```

Since I have to load AMDGPU for AMDGPU.functional() I would prefer if `__init__` would be quiet and the error message would be on "first use".

--- a/src/discovery/utils.jl
+++ b/src/discovery/utils.jl
@@ -10,6 +10,7 @@ function check_rocm_path(path::String)
     path2 = joinpath(path, "lib64")
     isfile(joinpath(path2, libname)) && @goto success
     return ""
+
     @label success
     @assert isdir(path2)
     return path2
@@ -19,8 +20,8 @@ end
 function find_roc_path()::String
     env_dir = get(ENV, "ROCM_PATH", "")
     if isdir(env_dir)
-    	rocm_path = check_rocm_path(env_dir)
-    	rocm_path != "" && return rocm_path
+        rocm_path = check_rocm_path(env_dir)
+        rocm_path != "" && return rocm_path
     end
 
     if Sys.islinux()
@@ -110,7 +111,7 @@ end
 
 function find_rocm_library(lib::String; rocm_path::String, ext::String = dlext)::String
     libdir = joinpath(rocm_path, rel_libdir)
-    @assert isdir(libdir)
+    isdir(libdir) || return ""
     for file in readdir(libdir; join=true)
         fname = basename(file)
         matched = startswith(fname, lib) && contains(fname, ext)
@@ -124,10 +125,10 @@ function find_ld_lld(rocm_path::String)::String
 
     dirs = (joinpath(rocm_path,"llvm", "bin"), joinpath(rocm_path,"bin"))
     hipconfig = Sys.which("hipconfig")
-     if !isnothing(hipconfig)
+    if !isnothing(hipconfig)
         clang_path = read(`$hipconfig --hipclangpath`, String)
         dirs = (dirs ..., clang_path)
-     end
+    end
     for dir in dirs
         exp_ld_path = joinpath(dir, lld_name)
         ispath(exp_ld_path) || continue

